codec/src/main/java/io/netty/handler/codec/DefaultTextHeaders.java

```diff
@@ -396,4 +401,188 @@ public TextHeaders clear() {
     private static Comparator<CharSequence> comparator(boolean ignoreCase) {
         return ignoreCase ? CHARSEQUENCE_CASE_INSENSITIVE_ORDER : CHARSEQUENCE_CASE_SENSITIVE_ORDER;
     }
+
+    /*
+     * This interface enables different implementations for adding/setting header values.
+     * Concrete implementations can control how values are added, for example to add all
+     * values for a header as a comma separated string instead of adding them as multiple
+     * headers with a single value.
+     */
+    private interface ValuesComposer {
+        TextHeaders add(CharSequence name, CharSequence value);
+        TextHeaders add(CharSequence name, CharSequence... values);
+        TextHeaders add(CharSequence name, Iterable<? extends CharSequence> values);
+
+        TextHeaders addObject(CharSequence name, Object value);
+        TextHeaders addObject(CharSequence name, Iterable<?> values);
+        TextHeaders addObject(CharSequence name, Object... values);
+
+        TextHeaders set(CharSequence name, CharSequence... values);
+        TextHeaders set(CharSequence name, Iterable<? extends CharSequence> values);
+
+        TextHeaders setObject(CharSequence name, Object... values);
+        TextHeaders setObject(CharSequence name, Iterable<?> values);
+    }
+
+    /*
+     * Will add multiple values for the same header as multiple separate headers.
+     */
+    private class MultipleFieldsValueComposer implements ValuesComposer {
+
+        @Override
+        public TextHeaders add(CharSequence name, CharSequence value) {
+            DefaultTextHeaders.super.add(name, value);
+            return DefaultTextHeaders.this;
+        }
+
+        @Override
+        public TextHeaders add(CharSequence name, CharSequence... values) {
+            DefaultTextHeaders.super.add(name, values);
+            return DefaultTextHeaders.this;
+        }
+
+        @Override
+        public TextHeaders add(CharSequence name, Iterable<? extends CharSequence> values) {
+            DefaultTextHeaders.super.add(name, values);
+            return DefaultTextHeaders.this;
+        }
+
+        @Override
+        public TextHeaders addObject(CharSequence name, Object value) {
```

Comment:
I think we can get rid of this.  See https://github.com/netty/netty/pull/3271/files#r22608659.